### ISSUE TITLE
[FIRParser] Locations for expressions are set wrong

### ISSUE THREAD

**Issue by lattner** [llvm/circt#1140]
We're blasting translateLocation in the FIR parser (it is about 6% of parser time).  Consider an example like this:

```
$ cat x.fir 
circuit foo :
  module foo:
    input mask_bit_2 : UInt<1>
    node mask_nbit_2 = eq(mask_bit_2, UInt<1>("h0")) @[Misc.scala 210:20]
```

which then parses like so:

```
$ firtool x.fir -disable-opt --mlir-print-debuginfo    
module  {
  firrtl.circuit "foo"  {
    firrtl.module @foo(in %mask_bit_2: !firrtl.uint<1>) {
      %c0_ui1 = firrtl.constant 0 : !firrtl.uint<1> loc(#loc3)
      %0 = firrtl.eq %mask_bit_2, %c0_ui1 : (!firrtl.uint<1>, !firrtl.uint<1>) -> !firrtl.uint<1> loc(#loc3)
      %mask_nbit_2 = firrtl.node %0  : !firrtl.uint<1> loc(#loc3)
    } loc(#loc2)
  } loc(#loc1)
} loc(#loc0)
#loc0 = loc("x.fir":0:0)
#loc1 = loc("x.fir":2:1)
#loc2 = loc("x.fir":3:3)
#loc3 = loc("Misc.scala":210:20)
```

Notice that the locator specified for the firrtl.constant and firrtl.eq nodes are *in the fir file itself* instead of being the locator specified on the statement.  This is incorrect (it won't be helpful to the user) and is inefficient in both time and memory (because we're creating garbage location nodes).

We should defer resolution of the SMLoc's until the statement is complete and apply a locator at that time (or not).

**Comment by lattner** [comment]
It's actually more like 12% of parser time, when you exclude verifier etc.

**Comment by lattner** [comment]
I'm working on this.

**Comment by lattner** [comment]
Fixed in 755f2af9
